src/Main.hs

```diff
@@ -132,45 +179,73 @@ prop lib = forAll arbitrary (ioProperty . test)
 runQC :: Serialize lib (BinTree Int) => lib -> IO ()
 runQC = quickCheckWith stdArgs{maxSuccess=1000} . prop
 
-generateBalancedTree :: Word32 -> IO (BinTree Int)
-generateBalancedTree 0 = Leaf <$> randomIO
-generateBalancedTree n = Tree <$> generateBalancedTree (n-1) <*> generateBalancedTree (n-1)
+generateBalancedTree = generateBalancedTree_ (generate $ arbitrary)
+generateBalancedTree_ r 0 = Leaf <$> r
+generateBalancedTree_ r n = Tree <$> generateBalancedTree_ r (n-1) <*> generateBalancedTree_ r (n-1)
 
 runBench :: IO ()
 runBench = do
-  tree <- force <$> generateBalancedTree 22
+ 
+  -- Data structures to (de)serialise
+  intTree <- ("BinTree Int",) . force <$> (generateBalancedTree 21 :: IO (BinTree Int))
+  directionTree <- ("BinTree Direction",) . force <$> (generateBalancedTree 21 :: IO (BinTree Direction))
+  carsDataset <- ("Cars dataset",) . force <$> carsData
+  let irisDataset = force ("Iris dataset",irisData)
+  -- let wordList = ("[Word8]",force [1..1000::Word8])
+  directionList <- ("[Direction]",) . force <$> mapM (\n -> generate $ arbitrary :: IO Direction) [1..1000000]
+
+  performMajorGC
 
-  bs_binary <- serialize Binary tree
-  bs_cereal <- serialize Cereal tree
-  bs_packman <- serialize Packman tree
-  bs_binary_cbor <- serialize CBOR tree
+  let jsonReport = "report.json"
+  let htmlReport = "report.html"
+  defaultMainWith (defaultConfig {jsonFile=Just jsonReport,reportFile=Just htmlReport})
+    $ benchs directionList ++ benchs carsDataset ++ benchs intTree ++ benchs directionTree ++ benchs irisDataset  -- ++ benchs wordList
 
-  putStrLn $ "Length of ByteString generated by binary:                " ++ show (BS.length bs_binary)
-  putStrLn $ "Length of ByteString generated by cereal:                " ++ show (BS.length bs_cereal)
-  putStrLn $ "Length of ByteString generated by packman:               " ++ show (BS.length bs_packman)
-  putStrLn $ "Length of ByteString generated by binary-serialise-cbor: " ++ show (BS.length bs_binary_cbor)
+  putStrLn "Summary:\n"
+  Right (_,_,reportsSummary -> reports) <- readJSONReports jsonReport
+  mapM_ (\(tst,ms) -> report tst "Time" "mSecs" ms) $ allTests reports
 
-  performMajorGC
+  sizes directionList
+  sizes directionTree
+  sizes intTree
+  sizes carsDataset
+  sizes irisDataset
+  -- sizes wordList
+
+sizes (name,obj) = mapM (\(n,s,d) -> (\ss -> (n,fromIntegral . BS.length $ ss)) <$> s obj) pkgs >>= report ("serialisation-"++name) "Size" "bytes"
+
+benchs (name,obj) = [
+  env (return obj) $ \sobj -> bgroup ("serialization-"++name) $ map (\(pkg,s,d) -> bench pkg (nfIO (s sobj))) pkgs
```

Comment:
No need to use `env` here as you're already forcing the datasets above (although you're doing it incorrectly, see my comment above). Or just remove `force` calls above and use `env` here. I'd pass dataset generator here and generate the dataset using `env`, but it's your call.